### PR TITLE
feat: re-resolve agentDefinitionKey and validate AgentDefinitionType on AgentInstance migration

### LINKED ISSUE #59093: Re-resolve agentDefinitionKey on AgentInstance migration
### 

### Description

On migration, the agent instance's associated agent definition is re-resolved to the target process version's definition. Element remaps resolve to the mapped target element's key.

Additionally, during process instance migration, the following constraint must be enforced: the `AgentDefinitionType` of the `AgentDefinition` associated with an `AgentInstance` must never change across a migration. Migrating from an element with no `AgentDefinition` to one with an `AgentDefinition` (or vice versa) is allowed — this covers switching element templates between one without the agent-definition marker and one with it. Migrating between two elements that both have an `AgentDefinition` is rejected if their `AgentDefinitionType` differs (e.g. `AI Agent Sub Process` -> `AI Agent Task`). The type comes from the process's `AgentDefinition` (type values: `AI Agent Sub Process`, `AI Agent Task`, `External Agent`), regardless of which element template is used.

### Motivation

After migration, the instance must reflect the definition of the version it now runs on — a new key per version. Migration must also ensure the `AgentDefinitionType` backing an `AgentInstance` never changes, regardless of which element template is targeted.

### Acceptance criteria

- [ ] `AgentInstance:MIGRATED` carries the target version's `agentDefinitionKey`.
- [ ] Element remaps resolve to the mapped target element's key.
- [ ] Migration is allowed when moving between an element with no `AgentDefinition` and one with an `AgentDefinition` (in either direction).
- [ ] Migration is rejected if both source and target elements have an `AgentDefinition` whose `AgentDefinitionType` differs.
- [ ] Tests cover migrate-with-remap and migrate-without-remap.
- [ ] Tests cover invalid migration scenarios (mismatched `AgentDefinitionType`).

### Links

- Feature epics: https://github.com/camunda/product-hub/issues/3703 (Queryable Agent registry), https://github.com/camunda/product-hub/issues/3702 (external-agent visibility)
- Parent epic: https://github.com/camunda/product-hub/issues/3462 (Real-Time Agent Visibility)
- Breakdown issue: #58439
- Parent: #59089

### Notes

- **SP:** 1
- **Depends on:** the create-time resolution task (adds `agentDefinitionKey` to the record). Soft: the migration-completion gate task (completes agent instances whose owning element is removed on migration), created separately.
- **Blocks:** —

### PR DESCRIPTION
**Pull request by mustafadagher**
## Description

When a process instance is migrated, `AgentInstance` records now correctly reflect the target process version through two complementary changes:

### 1. Re-resolving `agentDefinitionKey`

Every migrated `AgentInstance` is updated to point at the agent definition minted for the **target** process definition. This applies to:
- **Remapped elements**: the key resolves to the mapped target element's own agent definition.
- **Self-mapped elements**: the key is re-resolved to the target version's agent definition even when the element id is unchanged.

Previously, the `agentDefinitionKey` was left pointing at the *source* definition's agent definition after migration, breaking the link between the migrated agent instance and its new definition.

### 2. Validating `AgentDefinitionType` invariance

Migration is rejected (`INVALID_STATE`) when any `AgentInstance` would end up without an agent definition or with a different `AgentDefinitionType` after migration. Specifically:

- **No agent definition on the target element**: migration is rejected — an agent instance must always belong to an agent definition. This includes orphaned agent instances whose owning element was already dropped in the target process definition.
- **Different agent definition type**: migration is rejected when the source element's `AgentDefinitionType` (e.g. `AI_AGENT_TASK`) differs from the target element's type (e.g. `EXTERNAL_AGENT`). This complements `requireSameElementType` by distinguishing agent types that share a BPMN element type.
- **No agent instance on the source**: migration is *allowed* — elements with no existing agent instance can freely migrate onto an agent element (or vice versa).

The `AgentDefinitionType` is resolved correctly for both explicit `zeebe:agentDefinition` markers and the `zeebe:modelerTemplate` fallback (used by agents authored from element templates before the explicit marker existed).

### Validation scope

Validation runs as an upfront **per-agent-instance pass** over *all* agent instances of the process instance — active and orphaned alike — before the element-tree walk. This ensures orphaned instances (whose owning element already completed) are not silently migrated with an incompatible definition.

When the safety check is passed but the target element unexpectedly has no agent definition during re-resolution, a `SafetyCheckFailedException` is thrown (treated as a bug, not a user error).

## Related issues

Closes #59093

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationAgentInstanceBehavior.java [hunk 1]
```diff
@@ -73,6 +81,21 @@ final var record = agentInstanceState.getRecord(agentInstanceKey);
...
+   * Resolves the agent definition key the migrated agent instance should point at in the target
+   * process definition. Returns {@code -1} when the target element carries no agent definition,
+   * which is the allowed case of migrating an agent element onto a plain (non-agent) element.
+   */
+  private long resolveAgentDefinitionKey(
+      final DeployedProcess targetProcessDefinition, final String targetElementId) {
+    final var targetKey =
+        agentDefinitionState.getAgentDefinitionKey(
+            targetProcessDefinition.getKey(), BufferUtil.wrapString(targetElementId));
+    return targetKey == null ? UNSET_AGENT_DEFINITION_KEY : targetKey;
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationAgentInstanceBehavior.java]: ❌ An agent instance always belongs to an agent definition. We should not UNSET it. Instead, this is a SafetyCheckFailure. I suggest throwing a SafetyCheckFailedException.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java [hunk 2]
```diff
@@ -424,4 +425,216 @@ public void shouldMigrateAgentInstanceOfAdHocSubProcess() {
...
+                .getFirst()
+                .getValue())
+        .hasElementId("A")
+        .describedAs(
+            "agentDefinitionKey is re-resolved to the target definition's agent definition")
+        .hasAgentDefinitionKey(targetAgentDefinitionKey);
+  }
+
+  @Test
+  public void shouldResetAgentDefinitionKeyWhenTargetElementHasNoAgentDefinition() {
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java]: ❌ This should check that the migration is rejected. You're not allowed to migrate an agent element to a non-agent element, i.e. an agent instance must always have an agent definition.

File: qa/acceptance-tests/src/test/java/io/camunda/it/orchestration/AgentInstanceMigrationIT.java [hunk 3]
```diff
@@ -136,6 +138,57 @@ void shouldMigrateAgentInstanceOfAdHocSubProcess() {
     assertAgentInstanceMigratedTo(agentInstanceKey, targetProcess, targetElementId);
   }
 
+  @Test
+  void shouldRejectMigrationWhenAgentDefinitionTypeChanges() {
```
Comment (korthout) [inline qa/acceptance-tests/src/test/java/io/camunda/it/orchestration/AgentInstanceMigrationIT.java]: Nice to see that it works, but IMO we don't need an AT for this. They're slow, and this is an edge case that can be covered completely with EngineRule tests.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java [hunk 4]
```diff
@@ -637,4 +639,118 @@ private long agentDefinitionKey(final long processDefinitionKey, final String el
...
+        .hasIntent(ProcessInstanceMigrationIntent.MIGRATE)
+        .hasRejectionType(RejectionType.INVALID_STATE)
+        .hasRejectionReason(
+            String.format(
+                """
+                Expected to migrate process instance '%d' \
+                but active element with id 'A' has agent definition type 'AI_AGENT_TASK' \
+                while the mapped target element with id 'A2' has a different agent definition \
+                type 'EXTERNAL_AGENT'. \
+                Agent elements must be mapped to elements of the same agent definition type.""",
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java]: 👍 Great description!

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java [hunk 5]
```diff
@@ -637,4 +639,118 @@ private long agentDefinitionKey(final long processDefinitionKey, final String el
...
+                processInstanceKey));
+  }
+
+  @Test
+  public void shouldAllowMigrationFromNonAgentToAgentElement() {
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java]: ❌ This should only be allowed if there is no Agent Instance at this element yet. If there is no Agent Instance, then it doesn't break the "Each Agent Instance belongs to an Agent Definition" rule. But, an Agent Instance cannot be created without an Agent Definition, so the other case can't exist.

In other words, your test case below is fine, because it doesn't create an Agent Instance before migrating. But the test name does not reflect that.

💭 What instead should be possible is migration from a legacy one (detected using a fallback to the modeler template) as long as the type remains the same. But, looking at the commits ahead, you already cover all of those.

For example, `.zeebeModelerTemplate("io.camunda.connectors.agenticai.aiagent.jobworker.v1")` (fallback detection for AI Agent Sub Process) to the new dedicated agent definition one for AI Agent Sub Process`.zeebeAiAgentSubProcessDefinition()`.

Another example, `.zeebeModelerTemplate("io.camunda.connectors.agenticai.aiagent.v1")` (fallback detection for AI Agent Task) to the new dedicated agent definition one for AI Agent Task`.zeebeAiAgentTaskDefinition()`.

But not from AI Agent Task to AI Agent Sub Process, `.zeebeModelerTemplate("io.camunda.connectors.agenticai.aiagent.v1")` (fallback detection for AI Agent Task) to the new dedicated agent definition one for AI Agent Task`.zeebeAiAgentSubProcessDefinition()`, or vice-versa.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java [hunk 6]
```diff
@@ -694,10 +697,93 @@ public void shouldRejectMigrationWhenAgentDefinitionTypeChanges() {
...
+        .hasRejectionType(RejectionType.INVALID_STATE)
+        .hasRejectionReason(
+            String.format(
+                """
+                Expected to migrate process instance '%d' \
+                but the agent instance element with id 'A' has agent definition type 'AI_AGENT_TASK' \
                 while the mapped target element with id 'A2' has a different agent definition \
                 type 'EXTERNAL_AGENT'. \
-                Agent elements must be mapped to elements of the same agent definition type.""",
+                An agent instance's agent definition type must not change on migration.""",
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java]: 🔧 I liked the old wording better, but that's personal preference.

🔧 Alternatively: An agent instance's type must not change on migration

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java [hunk 7]
```diff
@@ -839,4 +846,135 @@ public void shouldAllowMigrationFromNonAgentToAgentElement() {
...
+                Expected to migrate process instance '%d' \
+                but the agent instance element with id 'A' has agent definition type 'AI_AGENT_TASK' \
+                while the mapped target element with id 'A2' has a different agent definition \
+                type 'EXTERNAL_AGENT'. \
+                An agent instance's agent definition type must not change on migration.""",
+                processInstanceKey));
+  }
+
+  @Test
+  public void shouldAllowMigrationBetweenTemplateAndMarkerOfSameAgentType() {
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java:910]: 👍 Exactly!

**Review by korthout — CHANGES_REQUESTED** [review]
Nice work @mustafadagher 👏 

👍 You wrote a bunch of good tests 

❌ There are a few cases that need another look. The two rules
1. An agent instance always belongs to an agent definition
2. An agent instance cannot change its type (i.e. the type of its agent definition)

🔧 I think the AT (IT) is not needed, it doesn't add anything that  can't be covered by EngineRule tests.

💭 I notice that you leave a lot of inline comments in the code. Please consider whether they earn their place.

**Comment by mustafadagher** [comment]
@korthout - I've fixed all your comments in 3ac81aa. Could you check again? 

💭 Once the changes satisfy the requirements, I'm thinking to squash all the commits from second till last into one commit before merging. ❓ WDYT?

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java [hunk 8]
```diff
@@ -198,34 +193,33 @@ public void shouldKeepStaleElementIdWhenOwningElementIsRemovedFromTargetProcessD
...
-                + "definition")
-        .hasProcessDefinitionKey(targetProcessDefinitionKey)
-        .hasBpmnProcessId(processId)
-        .hasProcessDefinitionVersion(2)
-        .hasVersionTag("v2")
-        .describedAs(
-            "elementId keeps its stale value \"A\" since no mapping instruction was provided and "
-                + "v2 of the process definition no longer contains an element with this id")
-        .hasElementId("A");
+            RecordingExporter.processInstanceMigrationRecords().onlyCommandRejections().getFirst())
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java]: 🔧 You can take the response of the .expectRejection().migrate() call to assert on. It's the same rejection record and avoids a second pass over the RecordingExporter's stream.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java [hunk 9]
```diff
@@ -624,18 +616,22 @@ public void shouldResetAgentDefinitionKeyWhenTargetElementHasNoAgentDefinition()
...
     Assertions.assertThat(
-            RecordingExporter.agentInstanceRecords(AgentInstanceIntent.MIGRATED)
-                .withRecordKey(agentInstanceKey)
-                .getFirst()
-                .getValue())
-        .hasElementId("A2")
-        .describedAs(
-            "agentDefinitionKey is reset because the target element has no agent definition")
-        .hasAgentDefinitionKey(-1L);
+            RecordingExporter.processInstanceMigrationRecords().onlyCommandRejections().getFirst())
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateAgentInstanceTest.java]: 🔧 Same here

**Review by korthout — APPROVED** [review]
Awesome 🚀 Great work @mustafadagher 

🔧 Found just a small test optimization that you could take along before merging.

LGTM 👍 

> I'm thinking to squash all the commits from second till last into one commit before merging

Squashing the entire diff into one commit would lose potentially meaningful information from the commits (the messages are scoped to their diff). I suggest to create new commits for the changes, or amend the old ones. You can use git absorb to amend fast. I use this in my CLAUDE.md to direct Claude to take the right action: 

>When responding to review feedback, decide between amend and new commit by semantic intent, not file overlap. Amend when the feedback refines or corrects the same intent the original commit expressed (typo, missed edge case, fix to the same bug). Create a new commit when the feedback changes what behavior is wanted — even if it edits the same files — because the original commit's intent stands and the new behavior is a separate semantic step.
